X-Opaque-ID header can cause "No colon found" in Hive

As of #1770 we are putting the hadoop job's name into the request as the X-Opaque-ID header for the sake of troubleshooting. Unfortunately Hive job names often contain a newline. When put into a header this is interpreted as a second header. And since there is no `:` you get an error like
```
2022-01-19T10:10:34.398-0600 [ERROR] [system.err] Caused by: org.elasticsearch.hadoop.rest.EsHadoopInvalidRequest: org.elasticsearch.hadoop.rest.EsHadoopRemoteException: illegal_argument_exception: No colon found
2022-01-19T10:10:34.398-0600 [ERROR] [system.err] null
2022-01-19T10:10:34.398-0600 [ERROR] [system.err]       at org.elasticsearch.hadoop.rest.RestClient.checkResponse(RestClient.java:487)
2022-01-19T10:10:34.398-0600 [ERROR] [system.err]       at org.elasticsearch.hadoop.rest.RestClient.execute(RestClient.java:444)
2022-01-19T10:10:34.398-0600 [ERROR] [system.err]       at org.elasticsearch.hadoop.rest.RestClient.execute(RestClient.java:438)
2022-01-19T10:10:34.398-0600 [ERROR] [system.err]       at org.elasticsearch.hadoop.rest.RestClient.execute(RestClient.java:406)
2022-01-19T10:10:34.398-0600 [ERROR] [system.err]       at org.elasticsearch.hadoop.rest.RestClient.mainInfo(RestClient.java:755)
2022-01-19T10:10:34.398-0600 [ERROR] [system.err]       at org.elasticsearch.hadoop.hive.EsStorageHandler.configureJobConf(EsStorageHandler.java:163)
2022-01-19T10:10:34.399-0600 [ERROR] [system.err]       ... 25 more
```
Hive queries could also have non-ASCII characters, which are invalid for http headers. This could also cause problems (unverified).

--- a/mr/src/main/java/org/elasticsearch/hadoop/cfg/Settings.java
+++ b/mr/src/main/java/org/elasticsearch/hadoop/cfg/Settings.java
@@ -768,7 +768,7 @@ public abstract class Settings {
     public abstract Properties asProperties();
 
     public Settings setOpaqueId(String opaqueId) {
-        setProperty(ES_NET_HTTP_HEADER_OPAQUE_ID, opaqueId);
+        setProperty(ES_NET_HTTP_HEADER_OPAQUE_ID, opaqueId.replace("\n", "").replace("\r", ""));
         return this;
     }
 

--- a/mr/src/main/java/org/elasticsearch/hadoop/cfg/Settings.java
+++ b/mr/src/main/java/org/elasticsearch/hadoop/cfg/Settings.java
@@ -768,8 +768,25 @@ public abstract class Settings {
     public abstract Properties asProperties();
 
     public Settings setOpaqueId(String opaqueId) {
-        setProperty(ES_NET_HTTP_HEADER_OPAQUE_ID, opaqueId.replace("\n", "").replace("\r", ""));
+        setProperty(ES_NET_HTTP_HEADER_OPAQUE_ID, cleanOpaqueId(opaqueId));
         return this;
+    }
+
+    /**
+     * Headers can't contain newlines or non-ascii characters. This method strips them out, returning whatever is left.
+     * @param opaqueId
+     * @return
+     */
+    private String cleanOpaqueId(String opaqueId) {
+        char[] chars = opaqueId.toCharArray();
+        StringBuilder cleanedOpaqueId = new StringBuilder(chars.length);
+        for (int i = 0; i < chars.length; i++) {
+            int character = chars[i];
+            if (character > 31 && character < 127) { //visible ascii
+                cleanedOpaqueId.append(chars[i]);
+            }
+        }
+        return cleanedOpaqueId.toString();
     }
 
     public String getOpaqueId() {

--- a/mr/src/test/java/org/elasticsearch/hadoop/cfg/SettingsTest.java
+++ b/mr/src/test/java/org/elasticsearch/hadoop/cfg/SettingsTest.java
@@ -1,0 +1,53 @@
+package org.elasticsearch.hadoop.cfg;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+
+import static org.junit.Assert.assertEquals;
+
+public class SettingsTest {
+    @Test
+    public void getXOpaqueId() throws Exception {
+        TestSettings testSettings = new TestSettings();
+        String opaqueId1 = "This is an opaque ID";
+        testSettings.setOpaqueId(opaqueId1);
+        assertEquals(opaqueId1, testSettings.getOpaqueId());
+        testSettings.setOpaqueId("This one\n has newlines\r\n and a carriage return");
+        assertEquals("This one has newlines and a carriage return", testSettings.getOpaqueId());
+        testSettings.setOpaqueId("This oñe has a non-ascii character");
+        assertEquals("This oe has a non-ascii character", testSettings.getOpaqueId());
+    }
+
+    public static class TestSettings extends Settings {
+        private Map<String, String> actualSettings = new HashMap();
+        @Override
+        public InputStream loadResource(String location) {
+            return null;
+        }
+
+        @Override
+        public Settings copy() {
+            return null;
+        }
+
+        @Override
+        public String getProperty(String name) {
+            return actualSettings.get(name);
+        }
+
+        @Override
+        public void setProperty(String name, String value) {
+            actualSettings.put(name, value);
+        }
+
+        @Override
+        public Properties asProperties() {
+            return null;
+        }
+    }
+}

